students/TimPauley/session06/good_perf_sample.py

```diff
@@ -0,0 +1,120 @@
+records and a module that reads that file.
+
+The module is badly written and probably can be made to run more quickly 
+and efficiently!
+
+You will look at the code and may immediately form a judgment about where 
+performance can be improved. Do not be tempted to make immediate changes!
+Be sure to use an evidence based approach, and show, with data, how you 
+were able to make improvements.
+
+Rewrite the module to improve performance. Provide evidence to demonstrate 
+the improvement.
+
+Your new module should be called good_perf.py, it should use identical 
+input and produce identical output to poor_perf.py
+
+'''
+
+"""
+Cleanup Version
+"""
+from timeit import timeit
+import datetime
+import csv
+
+
+def analyze(filename):
+    """
+    Program to read in a csv, loop over the rows in the file, 
+    convert the date from a string to a datetime object,
+    save count of years 2013-2018 in a dict, and count how many 
+    times "ao" appears in the file.
+    """
+
+    # timestamp for when we start executing the function
+    start = datetime.datetime.now()
+    with open(filename) as csvfile:
+        reader = csv.reader(csvfile, delimiter=',', quotechar='"')
+
+        # skip header
+        next(reader, None)
+        # we only care about dates after 01/01/2012, 
+        #this line converts that date to a datetime object 
+        #to be able
+        # to compare to other dates in the file
+        date =  datetime.datetime.strptime('01/01/2012', '%m/%d/%Y')
+        # save strptime as a variable that we can call on dates in the 
+        #file later to convert them to datetime objects
+        my_date = datetime.datetime.strptime
+
+        # counter for 'ao's in the file
+        found = 0
+        # counter dict for each year we care about counting in the file
+        year_count = {
+            "2013": 0,
+            "2014": 0,
+            "2015": 0,
+            "2016": 0,
+            "2017": 0,
+            "2018": 0
+        }
+
+        # iterate through 1,000,000 rows
+        for row in reader:
+            # store each date from the date column as a datetime object
+            row_date = my_date(row[5], '%m/%d/%Y')
+            # count each year 2013-2018 and save to year_count dictionary
+            if row_date > date:
+                if row_date.year == 2013:
+                    year_count["2013"] += 1
+                if row_date.year == 2014:
+                    year_count["2014"] += 1
+                if row_date.year == 2015:
+                    year_count["2015"] += 1
+                if row_date.year == 2016:
+                    year_count["2016"] += 1
+                if row_date.year == 2017:
+                    year_count["2017"] += 1
+                if row_date.year == 2018:
+                    year_count["2018"] += 1
+            # count number of 'ao's in text of column 6
+            if "ao" in row[6]:
+                found += 1
```

Comment:
You found the biggest time save, not going through the file twice. :) There are a few more optimizations that can be made, but they are pretty minor compared to the looping twice.